docs/concepts/time-and-windows-in-ksql-queries.rst

```diff
@@ -0,0 +1,305 @@
+
+Concrete implementations of timestamp extractors may retrieve or compute
+timestamps based on the actual contents of data records, like an embedded
+timestamp field, to provide event-time or ingestion-time semantics, or they
+may use any other approach, like returning the current wall-clock time at the
+time of processing to implement processing-time semantics.
+
+By creating a custom timestamp extractor class, you can enforce different notions
+or semantics of time, depending on the requirements of your business logic. For
+more information see :ref:`streams_developer-guide_timestamp-extractor`.
+
+Windows in KSQL Queries
+***********************
+
+Representing time consistently enables aggregation operations on streams and
+tables, like SUM, that have distinct time boundaries. In KSQL, these boundaries
+are named *time windows*.
+
+.. image:: ../img/ksql-window.png
+   :alt: Diagram showing the relationship between events and time in a KSQL stream
+
+Windowing lets you control how to group records that have the same key for
+stateful operations, like aggregations or joins, into time spans. KSQL tracks
+windows per record key.
+
+When using windows in your KSQL queries, aggregate functions are applied only
+to the records that occur within a specific time window. Records that arrive
+late are handled as you might expect: although the time window they belong to
+has expired, the late records are still associated with the correct window.
+
+You can specify a retention period for the window in your KSQL queries. This
+retention period controls how long KSQL waits for out-of-order or late-arriving
+records for a given window. If a record arrives after the retention period of a
+window has passed, the record is discarded and isn't processed in that window.
+
+.. Note for version 5.1
+    Starting in KSQL 5.1, a *grace period* configuration determines how long
+    to wait before closing a window. Retention time is still a valid parameter
+    that defines for how long the potentially closed window is stored. This
+    enables accessing the window with interactive queries, even if it's closed.
+
+In the real world, late-arriving records are always possible, and your KSQL
+applications must account for them properly. The system's time semantics
+determine how late records are handled. For processing-time, the semantics are
+“when the record is being processed”, which means that the notion of late records
+isn't applicable because, by definition, no record can be late.
+
+Late-arriving records are considered “late” only for event-time and ingestion-time
+semantics. In both cases, KSQL is able to handle late-arriving records properly.
+
+.. important::
+
+    KSQL is based on the Unix epoch time in the UTC timezone, and this can affect
+    time windows. For example, if you define a 24-hour tumbling time window, it will
+    be in the UTC timezone, which may not be appropriate if you want to have daily
+    windows in your timezone.
+
+Window Types

```

Comment:
For each of the three window types (tumbling, hopping, session), we should not just cover how to create the respective queries/tables, but also how to subsequently use the created tables, because this latter step is where a lot of users struggle.

- How does the data in a windowed table look like (notably the key)?
- How can I write a SELECT query (well, any query) on a windowed table?
- How can I access window-related information such as start/end time of the window?  (and e.g. ROWTIME vs. WINDOWSTART() and WINDOWEND() times)

Reply:
I opened KSQL-1931 to track this.